### DISCUSSION TITLE
Switchover in ectd election API

### DISCUSSION THREAD

**Discussion by MatteoGioioso** [etcd-io/etcd#14716] [Q&A]
I am writing a HA solution for a database and I am using the go client election API: https://pkg.go.dev/go.etcd.io/etcd/client/v3/concurrency#Election

This is the pseudo-go-code of what I have at the moment:

```

import (
   clientv3 "go.etcd.io/etcd/client/v3"
   "go.etcd.io/etcd/client/v3/concurrency"
   ....
)

func main {
  ...
  id := uuid.V4()
  e := concurrency.NewElection(my_session, "/leader/")
  go e.Campaign(ctx, id)
  
  for {
    leader, _ := e.Leader(ctx)
    if leader.Kvs[0].Value == id {
      print("I am the leader")
    } else {
      print("I am the replica")
    }

    time.Sleep(30*time.Seconds)
  }
}

``` 

This works fine, if I start multiple processes of the same application and stop the leader, another one will be elected automatically.

Now I would like to add the possibility of switchover. They idea would be to add an endpoint and send the id of the instance I want to promote. 

The client election API contains a method called `Proclaim`, which I think it does what I need, however it just change the value of the leader key.

There is also another method `ResumeElection`, the signature seems to suggest that I can assign an arbitrary leader key, however is not explained how it works and also it may require all the applications to recreate a session or a new election object which sounds quite cumbersome.

Is there anything I am missing? Is there a simple way to promote a specific replica to leader?
Thanks

**Comment by shreemaan-abhishek** [comment]
You can use `election.Resign()` to make a given instance lose the leadership status. This would cause other instances to participate in the leader election. Is this what you are looking for?

Btw I am writing a leader election client based on etcd that would provide a switchover out of the box. Check [this](https://github.com/shreemaan-abhishek/rasputin) out.

**Reply by MatteoGioioso** [reply]
Yes, I want a given instance to lose the leadership status, but at the same time elect a specific instance as a leader. Not sure if you ever heard about Patroni: https://patroni.readthedocs.io/en/latest/rest_api.html#switchover-and-failover-endpoints
It has an API call in which you give a specific node identifier and it will try to elect it as a leader. 
It does use etcd as DCS.

**Reply by shreemaan-abhishek** [reply]
@MatteoGioioso did you take a look at session.WithLease()?
<img width="476" alt="image" src="https://user-images.githubusercontent.com/61597896/201459628-bb575e65-41e6-4bea-bd84-c115ec71fd4e.png">

**Reply by MatteoGioioso** [reply]
Thanks. No I did not, but from the description doesn't look like it can help me.

**Reply by shreemaan-abhishek** [reply]
hmm, let me know if you find something.

**Reply by MatteoGioioso** [reply]
I will try to play a bit around, but I am starting to think that is a missing feature of the SDK. Thanks!
